### ISSUE TITLE
作者请问下train_distill_reason在注意力中没有使用attention_mask

### ISSUE THREAD

**Issue by yudaiyan** [jingyaogong/minimind#520]
<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/d6c0fdfb-a800-4440-8e14-b4699fffffb8" />

如图，我验证过，没有使用attention_mask，训练distill_reason，推理可以得到正常的<think><answer>

<img width="1535" height="779" alt="Image" src="https://github.com/user-attachments/assets/2bbc9936-7160-4872-aab5-fbaf73453be3" />

使用attention_mask作为填充掩蔽加到extended_attention_mask训练，推理得到的是sft的结果，就没有<think><answer>。

我是从full_sft训练的。

<img width="1484" height="649" alt="Image" src="https://github.com/user-attachments/assets/9fbcb8f6-9d96-4ffe-bf23-db8b73c8a6ca" />

我问ai都是要加填充掩蔽的。

请教下作者，为什么不加attention_mask可以训练出reason，加了填充掩蔽反而不行？

**Comment by yudaiyan** [comment]
<img width="1928" height="1042" alt="Image" src="https://github.com/user-attachments/assets/b3041bd6-dcce-41ec-84d6-ed64a1f187bc" />
这里attention_mask=None，代表 padding的token 会参与注意力计算。

**Comment by jingyaogong** [comment]
“使用attention_mask作为填充掩蔽加到extended_attention_mask训练，推理得到的是sft的结果，就没有。”
我没理解你这句话的意思，抛开这句回答其他的。

---

“这里attention_mask=None，代表 padding的token 会参与注意力计算。”
LLM是个causal model，右padding不影响左边tokens的注意力计算。真实token不会看到右侧的PAD（因为causal mask），与左padding不同（左padding会被真实token看到），左padding确实需要attention_mask，但是这里为了简化没必要加attn_mask。
非要说影响就是浪费计算资源，pad在序列占比越大越浪费，但这里数据很明显pad不大。另一个可能的影响是如果用Layer Norm 在整个序列上计算统计量，Pad会影响均值方差，然而这里用的是RMSNorm，所以也不受影响。

至于你说传attention_mask进来会影响行为。右padding的attention_mask形如 `[1 1 1 1 1 1 1 0 0 0]` 不可能改变行为，如果改变了，那么请检查你的attention_mask的误用情况。

**Comment by yudaiyan** [comment]
@jingyaogong 
我描述下具体流程，我基于full_sft蒸馏，训练100个样本，推理如下：

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/481e6b8f-4059-425a-8b99-d062b64a7726" />

我进行了以下改动，基于full_sft蒸馏，训练100个样本，推理如下：

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/e30c8345-8638-4934-80e8-fcc998646122" />

我改动的那点，导致无法训练出reason。

**Comment by jingyaogong** [comment]
@yudaiyan `loss_mask`和`attention_mask`不是一个东西，这是严重误解

lossmask的user和system消息都是0掩码，前向计算attention如果把这种mask传进去，没法正确训练。

**Comment by jingyaogong** [comment]
正确的attention_mask是：内容部分全1，pad部分=0（对于右pad构造和使用这种mask几乎没意义）
这和loss_mask是两个概念
